pkg/ingress/ingress.go

```diff
@@ -0,0 +1,178 @@
+package ingress
+
+import (
+	"context"
+
+	"github.com/spf13/viper"
+	extv1beta "k8s.io/api/extensions/v1beta1"
+	netv1beta "k8s.io/api/networking/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ExtensionAPI indicate k8s networking.k8s.io/v1beta1 should be used on the current cluster
+const ExtensionAPI = "extension"
+
+// NetworkingAPI indicate k8s extensions/v1beta1 should be used on the current cluster
+const NetworkingAPI = "networking"
+
+// Client wrap around k8s client, and decide which ingress API should be used, depending on cluster capabilities.
+type Client struct {
+	client  client.Client
+	rClient client.Reader
+}
+
+// NewIngressClient Creates a new Ingress.client wrapper.
+func NewIngressClient(client client.Client, reader client.Reader) *Client {
+	return &Client{
+		client:  client,
+		rClient: reader,
+	}
+}
+
+func (c *Client) fromExtToNet(ingress extv1beta.Ingress) netv1beta.Ingress {
+	oldIngress := netv1beta.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Ingress",
+			APIVersion: "extensions/v1beta1",
+		},
+		ObjectMeta: ingress.ObjectMeta,
+		Spec: netv1beta.IngressSpec{
+			Backend: &netv1beta.IngressBackend{
```

Comment:
Whet if backend is not defined? In the `fromNetToExt` `ingress.Spec.Backend` is checked to see if nil.